### DISCUSSION TITLE
OpenTK 5 pre11 - What is libgio?

### DISCUSSION THREAD

**Discussion by isolin** [opentk/opentk#1746] [Q&A] [answered]
I would like to try out the preview of Vulkan bindings. So following the [PAL2 intro](https://nogginbops.github.io/opentk.net/learn/pal2/Introduction.html) I created a new console app and tried to run `Toolkit.Init(options)`. However, that fails with
```
Exception has occurred: CLR/System.EntryPointNotFoundException
An unhandled exception of type 'System.EntryPointNotFoundException' occurred in OpenTK.Platform.dll: 'Unable to find an entry point named 'g_power_profile_monitor_dup_default' in shared library 'Gio'.'
   at OpenTK.Platform.Native.LibGio.g_power_profile_monitor_dup_default()
   at OpenTK.Platform.Native.X11.X11ShellComponent.Initialize(ToolkitOptions options)
   at OpenTK.Platform.Toolkit.Init(ToolkitOptions options)
   at Program.<Main>$(String[] args) in /home/ubuntu/Code/VulkanTKTest/Program.cs:line 15
```

I am on Ubuntu 20.04 with `gtk-3-dev` installed. Is there any package I should add for `libgio` to work?

Also a second question - is there a complete Vulkan init exapmle yet? E.g. drawing a triangle. EDIT: I found [this one](https://github.com/opentk/opentk/blob/opentk5.0/tests/VulkanTestProject/Program.cs) so far.

**Comment by NogginBops** [comment]
To provide theme events and the current power saver state we use `libgio` that is part of the GTK project.
I'm not sure why it's not available by default (we might need to look at alternative solutions to libgio in that case...), but I think the "easy" solution would be to install `libgio-2.0-dev`, through e.g. `sudo apt install libgio-2.0-dev` or similar.

Looking at the debian apt repository it seems like the `libgio` is only a suggested package as part of ` libglib2.0-dev` and required not a dependency of it like I would have expected.

As for Vulkan examples the example that clears in rainbow colors is the only Vulkan sample we have available atm. I aim to have a more complete version of that Vulkan test program, but I haven't gotten to it yet.

**Reply by isolin** [reply]
There is no such package in bionic 20.04. I installed `libgio2.0-cil-dev` but it did not help. I will try with 24,04.

UPDATE: I tried in a 24.04 container. It's worse:
```
Exception thrown: 'System.NotSupportedException' in OpenTK.Platform.dll: 'This platform doesn't support the OpenTK.Platform.ISurfaceComponent component yet.'
   at OpenTK.Platform.Native.PlatformComponents.GetPlatformComponent[TComp](PalComponents comp
```

**Reply by NogginBops** [reply]
> `Exception thrown: 'System.NotSupportedException' in OpenTK.Platform.dll: 'This platform doesn't support the OpenTK.Platform.ISurfaceComponent component yet.'`

This exception is being caught internally, you likely have it so that the debugger breaks at every exception that is thrown even if the exception itself will be caught.

I booted up my linux machine to figure out what package provided `libgio2.0.so` and it turns out it's `libglib2.0-dev` and not `libgio-2.0-dev` (which does not exist).

Can you try installing `libglib2.0-dev` and see if this fixes the issue?

**Reply by isolin** [reply]
True, I had all exceptions on. 

Here my findings regarding Gio:

- `libglib2.0-dev` is already installed on my 20.04 machine and it has been installed already when the error occurred for the first time
- `libglib2.0-dev` is also installed in my 24.04 testing container, but that one did not complain about Gio, I just had to install `sudo apt install zenity` and it runs fine now.

So maybe there are some other issues regarding `Gio` on Ubuntu 20.04 that are preventing it to load, since it is actually present.

**Reply by NogginBops** [reply]
> So maybe there are some other issues regarding Gio on Ubuntu 20.04 that are preventing it to load, since it is actually present.

Are you able to find the `libgio.so` file in your 20.04 installation? It could be that the file on 20.04 has a different postfix than we are looking for atm. The file should be located at something like `/usr/lib/x86_64-linux-gnu/libgio-2.0.so` but it may be called `libgio-2.0.so.1` or similar.

**Reply by NogginBops** [reply]
Here is the relevant list of names we are looking for, so if your binary name doesn't match any of these we have to add more names to the list.
https://github.com/opentk/opentk/blob/opentk5.0/src/OpenTK.Platform/Native/DllResolver.cs#L120-L124

**Comment by isolin** [comment, marked as answer]
There is `/usr/lib/x86_64-linux-gnu/libgio-2.0.so` in the 20.04 installation. The error message says it is missing an endpoint.

Focal has [v2.64.6-1](https://launchpad.net/ubuntu/focal/+package/libglib2.0-dev)
Jammy has [v2.72.4](https://launchpad.net/ubuntu/jammy/+package/libglib2.0-dev)
Noble has [v2.80.0](https://launchpad.net/ubuntu/noble/+package/libglib2.0-dev)

Looking at glib source I found the [commit](https://github.com/GNOME/glib/commit/889bdb994fed44344a84ad01aa5633a1b1b62b19) which added `the g_power_monitor_dup_default` and it was in `v 2.69.1` in July 2021 which is some time between Focal and Jammy. So Jammy's `libgio` has the endpoint, but Focal's doesn't. This means, that all OpenTK's components that require it are not compatible with Linux versions released before July 2021. Perhaps one could upgrade the library manually, but I wouldn't dare given it is quite low level.

**Reply by NogginBops** [reply]
Wow! Thanks for finding the commit in glib that added this functionality.
When I was developing this I was under the impression that the functions I where using had been a part of glib for some time, but apparently not...

From the looks of things it seems like the power monitor implementation isn't hard and that it might be possible to implement a fallback implementation that basically copies the glib implementation.
Alternatively we could just make the code not send power saver notifications if that function is not yet available.
